netbench-orchestrator/src/ec2_utils.rs

```diff
@@ -1,21 +1,193 @@
+            .collect();
+
+        ec2_client
+            .terminate_instances()
+            .set_instance_ids(Some(ids))
+            .send()
+            .await
+            .map_err(|err| OrchError::Ec2 {
+                dbg: err.to_string(),
+            })?;
+
+        Ok(())
+    }
+
+    // Attempt to delete the Security Group.
+    //
+    // Retry the operation if the resource is still 'in-use' (`DependencyViolation`). Since an
+    // EC2 instance takes time to fully terminate, a Security Group could be 'in-use' until the
+    // EC2 host is fully cleaned up.
+    async fn delete_security_group(&self, ec2_client: &aws_sdk_ec2::Client) -> OrchResult<()> {
+        info!("Start: deleting security groups");
+
+        let mut deleted_sec_group = ec2_client
+            .delete_security_group()
+            .group_id(self.security_group_id.to_string())
+            .send()
+            .await;
+
+        let mut retries = RETRY_COUNT;
+        while deleted_sec_group.is_err() && retries > 0 {
+            let err = deleted_sec_group.as_ref().expect_err("expect error");
+            debug!(
+                "deleting security group. retry {retries}. \nerror: {:?}",
+                err
+            );
+
+            if let Some(service_err) = err.as_service_error() {
+                if service_err.meta().code() == Some("DependencyViolation") {
+                    tokio::time::sleep(RETRY_BACKOFF).await;
+                    deleted_sec_group = ec2_client
+                        .delete_security_group()
+                        .group_id(self.security_group_id.to_string())
+                        .send()
+                        .await;
+                } else {
+                    break;
+                }
+            } else {
+                break;
+            }
+
+            retries -= 1;
+        }
+
+        deleted_sec_group.map_err(|err| {
+            error!("abort deleting security group {}", self.security_group_id);
+            OrchError::Ec2 {
+                dbg: err.to_string(),
+            }
+        })?;
+
+        Ok(())
+    }
+
+    // Attempt to delete the Placement Group.
+    //
+    // Retry the operation if the resource is still 'in-use' (`InvalidPlacementGroup.InUse`). Since
+    // an EC2 instance takes time to fully terminate, a Placement Group could be 'in-use' until the
+    // EC2 host is fully cleaned up.
+    async fn delete_placement_group(&self, ec2_client: &aws_sdk_ec2::Client) -> OrchResult<()> {
+        info!("Start: deleting placement groups");
+
+        for (_az, placement_group) in self.placement_map.iter() {
+            let mut retries = RETRY_COUNT;
+
+            let placement_group_name = placement_group.group_name().ok_or(OrchError::Ec2 {
+                dbg: "Failed to get placement_group name".to_string(),
+            })?;
+            debug!("Start: deleting placement group: {placement_group_name}");
+
+            let mut delete_placement_group = ec2_client
```

Comment:
For all of these conditions, would it be neater to do it inside of an explicit `loop`? That way you wouldn't have the copy-pasted api request.

```
while retries < MAX_ALLOWED_RETRIES {
    let result = api_request()

    if DependencyViolation {
        sleep()
    }
}
```

And then can we also add an error log for if we stopped retrying and gave up?